1.1-RC.2 Pilot SIGSEGV under load

build: `gcr.io/istio-release/pilot:release-1.1-20190305-09-16`

Pilot was under load  (limit 6vCPUs, using 5 of those) but under stead state.
I observed a restart 3 times.

```
8,Name:svc38-0-3-b65b57b47-qkkqc,UID:ebe484e8-3ef4-11e9-bc04-42010a80005e,APIVersion:,ResourceVersion:20998480,FieldPath:,}}] [{http-web 8080 TCP}]}] [0xc634184000]
2019-03-06T18:07:21.052516Z	info	Handling event update for pod svc10-0v1-78fb965bb7-5gbp9 in namespace service-graph10 -> 10.16.55.229
2019-03-06T18:07:21.052541Z	info	Handling event update for pod svc34-0v1-868b8475f9-vf4mz in namespace service-graph34 -> 10.16.32.43
2019-03-06T18:07:21.052573Z	info	Handling event update for pod svc17-0-1-0-54884896c5-bflkd in namespace service-graph17 -> 10.16.47.247
2019-03-06T18:07:21.052596Z	info	Handling event update for pod svc09-0-4-56467bf559-klnd8 in namespace service-graph09 -> 10.16.13.23
2019-03-06T18:07:21.052680Z	warn	Omitting listener for management address 10.16.36.51_8080 (socket_address:<address:"10.16.36.51" port_value:8080 > ) due to collision with service listener 10.16.36.51_8080 (socket_address:<address:"10.16.36.51" port_value:8080 > )
2019-03-06T18:07:20.961021Z	warn	Omitting listener for management address 10.16.5.254_8080 (socket_address:<address:"10.16.5.254" port_value:8080 > ) due to collision with service listener 10.16.5.254_8080 (socket_address:<address:"10.16.5.254" port_value:8080 > )
2019-03-06T18:07:21.055371Z	info	ads	CDS: PUSH 2019-03-06T18:07:19Z/21 for sidecar~10.16.38.171~svc32-0-2-0-6467dd6fb7-rrtd5.service-graph32~service-graph32.svc.cluster.local-693 "10.16.38.171:38622", Clusters: 1856, Services 1596
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x16e0077]

goroutine 46256 [running]:
istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).updateServiceShards(0xc420617130, 0xc5cf71ed00, 0xc5cf71ed00, 0x0)
	/workspace/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/eds.go:375 +0x997
istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).Push(0xc420617130, 0x1, 0xc5cdada030)
	/workspace/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/discovery.go:329 +0x37e
istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).doPush(0xc420617130, 0x1d6ed01)
	/workspace/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/discovery.go:384 +0xa0
istio.io/istio/pilot/pkg/proxy/envoy/v2.(*DiscoveryServer).(istio.io/istio/pilot/pkg/proxy/envoy/v2.doPush)-fm(0x1)
	/workspace/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/discovery.go:402 +0x33
created by istio.io/istio/pilot/pkg/proxy/envoy/v2.handleUpdates
	/workspace/go/src/istio.io/istio/pilot/pkg/proxy/envoy/v2/discovery.go:446 +0x426
```

--- a/pilot/pkg/proxy/envoy/v2/eds.go
+++ b/pilot/pkg/proxy/envoy/v2/eds.go
@@ -371,8 +371,17 @@ func (s *DiscoveryServer) updateServiceShards(push *model.PushContext) error {
 
 	s.mutex.Lock()
 	for k, v := range svc2account {
-		ep := s.EndpointShardsByService[k]
-		ep.ServiceAccounts = v
+		if ep := s.EndpointShardsByService[k]; ep != nil {
+			ep.ServiceAccounts = v
+			continue
+		}
+		// we have not seen this service before.
+		// We will let edsUpdate() add endpoints to the list.
+		// Just record service accounts here.
+		s.EndpointShardsByService[k] = &EndpointShards{
+			Shards:          map[string][]*model.IstioEndpoint{},
+			ServiceAccounts: v,
+		}
 	}
 	s.mutex.Unlock()
 

